Process Plugin

e1528532 opened the pull request:
# Purpose

Implements a plugin which uses libpluginprocess to encapsulate arbitrary plugins in an own process

# Checklist

- [X] commit messages are fine (with references to issues)
- [X] I added unit tests
- [X] I ran all tests and everything went fine
- [X] affected documentation is fixed
- [X] I added code comments, logging, and assertions (see doc/CODING.md)
- [X] meta data is updated (e.g. README.md of plugins)
- [X] release notes are updated (doc/news/_preparation_next_release.md)

@markus2330 as promised

e1528532 commented:
jenkins build all please

e1528532 commented:
jenkins build all please

e1528532 commented:
interesting, output_warnings crashed in case there were multiple warnings returned in a test. However this fix was so tiny and this was apparently the first plugin that made use of that in a test, i just did it here as i think after that this PR is fine too.

e1528532 commented:
Now that invoke has error messages i see the reason why this one test fails, apparently it fails to resolve some asan symbol correctly. Maybe travis doesn't set up the symbolizer paths appropriately as described in the documentation. Lets see if this works on our own build server.

e1528532 commented:
jenkins build clang-asan please

e1528532 commented:
Unfortunately, the issue [still seems to persist](https://travis-ci.org/ElektraInitiative/libelektra/jobs/338590555):

```
reason: of module: libelektra-dump.so, because: /home/travis/build/ElektraInitiative/build/lib/libelektra-dump.so: undefined symbol: __ubsan_vptr_type_cache
```

I cannot reproduce this with my local debian though. The only idea that i'd still have is that this LD_PRELOAD thing is required as well for that one build, though it'd be strange as all the other tests using asan work fine as well. But the error message looks like the asan library is missing to me..

>LD_PRELOAD=/usr/lib/clang/3.8.0/lib/linux/libclang_rt.asan-x86_64.so run_asan
>or on Debian:
>LD_PRELOAD=/usr/lib/llvm-3.8/lib/clang/3.8.1/lib/linux/libclang_rt.asan-x86_64.so run_asan

However the travis job seems to use version 5 of clang. Is there any way i can find out the path of this library on that one travis machine somehow, without blindly trying around until it fits @sanssecours (you seem to know a lot about travis, so maybe you know this too) @markus2330 ?

markus2330 commented:
You could run a `find / -name '*libclang_rt.asan-x86_64*'` or similar on the travis server.

Wouldn't it be the better approach to search for this library using CMake? CMake searches in various standard folders by default, so we have a much better chance that it works on new OSs.

e1528532 commented:
> Wouldn't it be the better approach to search for this library using CMake

On the long run it most likely is, but that'd be worth an own PR i guess and to be honest i want to focus more on my type checker to move forward there, so i'd be happy with an LD_PRELOAD approach for now ;)

markus2330 commented on .travis.yml:
```diff
@@ -173,12 +173,13 @@ before_script:
       ninja run_all
     elif [[ "$ASAN" = "ON" && "$TRAVIS_OS_NAME" == "linux" ]]; then
-      LD_PRELOAD=$(find /usr/ -name '*libclang_rt.asan-x86_64.so') ASAN_OPTIONS=symbolize=1 ASAN_SYMBOLIZER_PATH=$(which llvm-symbolizer) ninja run_all
+      LD_PRELOAD=$(find /local/mnt/workspace/tmp/ -name '*libclang_rt.asan-x86_64.so') ASAN_OPTIONS=symbolize=1 ASAN_SYMBOLIZER_PATH=$(which llvm-symbolizer) ninja run_all
```
Simply use `find /` or does it take so long? @sanssecours what do you think?

e1528532 replied:
the thing is the asan library in there doesn't work it says linked to incompatible asan runtime there and points me to some strange path /local/mnt/workspace/tmp which i am now hackily investigating what it contains for us ;)

e1528532 replied:
ok apparently that path doesn't even exist and asan doesn't like the runtime in /usr/ although there are only 5.0.0 runtimes and i have no idea where else it magically picks it up from. lets just search / then.

e1528532 replied:
The search yielded nothing new, there seems to be only /usr/local/clang-5.0.0/lib/clang/5.0.0/lib/linux/libclang_rt.asan-x86_64.so but when i LD_PRELOAD on that it complains its  not the asan runtime linked against...

```
clang version 5.0.0 (tags/RELEASE_500/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/clang-5.0.0/bin
```

this would also indicate that the runtime in local/clang/... should be correct so i'm not sure how to continue. I'm tempted to just exclude the process plugin in that constellation as we have our own asan build anyway. Its not executed yet though so lets see if that works as a replacement.

markus2330 replied:
Yes, it is difficult to find the correct one afterwards. I still suggest to do it via CMake, e.g., did you look at https://github.com/arsenm/sanitizers-cmake

e1528532 replied:
while this seems like a fine idea, in my experience changing such build-related things usually turn out to be a lot of work even if it only looks like changing a few simple line. So I'd prefer to follow another approach for this PR for now (like blacklisting it for this one travis asan build in case it works on our own asan build). We can open an issue about using such a cmake approach for the sanitizer for the future.

e1528532 commented:
@markus2330 https://build.libelektra.org/jenkins/job/elektra-clang-asan/393/ looks like our own asan build works fine and only this one linux config on travis is affected. If its ok, I'd disable the travis asan test for the process plugin for now until we've fixed the asan configuration via cmake or something like that.

sanssecours commented:
Looks like we need to preload [`libclang_rt.ubsan_standalone_cxx-x86_64.so`](https://travis-ci.org/sanssecours/elektra/builds/338933829) instead of `libclang_rt.asan-x86_64.so`. I updated the PR accordingly in my copy of the repository.

@e1528532 🙏 Please wait until I update this PR.

sanssecours commented:
I just pushed an update containing my changes. I also dropped and merged some of the latest commits using `git rebase --interactive`. If you think the updates are too intrusive Armin, then please feel free to just overwrite them.

> @sanssecours what do you think?

Preloading a library via an environment variable does not sound like a good idea. Hopefully we find a better solution for this problem in the future.

e1528532 commented:
> I just pushed an update containing my changes. I also dropped and merged some of the latest commits using git rebase --interactive

very nice! i wasn't even aware that you can get rid of such "try&error" style commits after you've pushed them easily. its much cleaner now without the unnecessary commits, so thanks for that.

> Hopefully we find a better solution for this problem in the future.

yes we should evaluate the cmake asan integration - maybe that would take care about such issues.

However, lets see now if the tests will pass this time. It'd be great to see this merged!

e1528532 commented:
jenkins build all please

e1528532 commented:
thanks to @sanssecours it finally looks good now. @markus2330 i think we can merge this first version now., its still experimental anyway i haven't tried too much with it yet apart from the simple test.

markus2330 commented on src/libs/invoke/invoke.c:
```diff
@@ -264,14 +264,16 @@ int elektraInvoke2Args (ElektraInvokeHandle * handle, const char * elektraPlugin
 
-	typedef int (*elektra2Args) (Plugin *, KeySet *, Key *);
-	elektra2Args func = *(elektra2Args *)elektraInvokeGetFunction (handle, elektraPluginFunctionName);
+	// If we cast this right away it will hang if the function is not
```
sentence?

markus2330 commented on src/plugins/process/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+
+find_package (Pluginprocess)
+
+if (PLUGINPROCESS_FOUND)
```
Isn't dump needed, too?

e1528532 replied:
yes thats what i thought infos/needs takes care about, will correct it

markus2330 commented on src/plugins/process/README.md:
```diff
@@ -0,0 +1,53 @@
+- infos = Information about the process plugin is in keys below
+- infos/author = e1528532 <e1528532@libelektra.org>
```
Your name please.

markus2330 commented on src/plugins/process/README.md:
```diff
@@ -0,0 +1,53 @@
+- infos = Information about the process plugin is in keys below
+- infos/author = e1528532 <e1528532@libelektra.org>
+- infos/licence = BSD
+- infos/needs = dump
```
needs tells that dump should be loaded as separate plugin, this is not what we want

markus2330 commented on src/plugins/process/README.md:
````diff
@@ -0,0 +1,53 @@
+
+```sh
+# Mount the dump plugin a separate process via the process plugin to the cascading namespace `/examples/process`
+sudo kdb mount config.file /examples/process process plugin=dump
````
Does not work for me, I get the error:
```
The command bin/kdb mount terminated unsuccessfully with the info:
Was not able to load such a plugin!

Maybe you misspelled it, there is no such plugin or the loader has problems.
You might want to try to set LD_LIBRARY_PATH, use kdb-full or kdb-static.
Errors/Warnings during loading were:

Please report the issue at https://issues.libelektra.org/
```

Even "kdb check process" fails.


Please also add this shell recorder test in tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt then hopefully the build server will tell you the same problem I have locally.

e1528532 replied:
it is expected that kdb check process fails, as without a plugin configuration it does not work because it doesn't know what plugin shall be proxied.

e1528532 replied:
@markus2330 regarding the mounting issue, thats strange, i am pretty sure this worked at some point but i can reproduce it also. Again, logging reveals the actual cause (I've added a simple logging statement for debugging just after it gets the plugin configuration like
```
KeySet * processConfig = elektraPluginGetConfig (handle);
Key * pluginName = ksLookupByName (processConfig, "/plugin", KDB_O_NONE);
ELEKTRA_LOG ("we are trying to open the plugin %s", keyString (pluginName));
```
)

but it looks like the configuration parameter is not present? Other plugins using config seem to handle it the same way though, so i'm not exactly sure yet where the error hides.

```
sudo kdb mount test.dump user/examples/process process plugin=dump
...
src/plugins/process/process.c:62:elektraProcessOpen: we are trying to open the plugin (null)
```

markus2330 replied:
You should definitely always add warnings in such situations. It should never be required to enable logging to debug wrong plugin parameters.

> which would have complained about the missing name?

Where do you implement this check? Do you also pass the checkConfig calls to the proxied plugins?

e1528532 replied:
> You should definitely always add warnings in such situations.

done

> Where do you implement this check?

Currently it was only in the checkConfig function but now i've also included the same check in the startup function so users get the warning in any case. the little overhead should be acceptable for that.

I don't pass checkConfig calls to proxied plugins though, as there is no Plugin handle in the checkConfig function signature so i can't really access the child process in there.

markus2330 commented on src/plugins/process/process.c:
```diff
@@ -0,0 +1,243 @@
+	// If we cast this right away although the function wasn't found it will cause a deadlock
+	const void * rawFunc = elektraInvokeGetFunction (handle, elektraPluginFunctionName);
+
+	if (!rawFunc) return -2;
```
No error set?

e1528532 replied:
no it just crashed right away and when i separated the cast from getting the pointer it worked - so that was more fixed by trial and error ;)

markus2330 commented on src/plugins/process/process.c:
```diff
@@ -0,0 +1,243 @@
+	return elektraPluginExport ("process",
+		ELEKTRA_PLUGIN_OPEN,	&elektraProcessOpen,
+		ELEKTRA_PLUGIN_CLOSE,	&elektraProcessClose,
+		ELEKTRA_PLUGIN_GET,		&elektraProcessGet,
```
formatting?

markus2330 replied:
still not resolved

e1528532 replied:
funny, on my system it looks normal this way, but ok deleted one tab ;)

markus2330 replied:
Hopefully GitHub did not change their tab setting (or is different than ours)... I trusted the GitHub view and did not check locally. Our settings are in .clang-format: TabWidth is 8

markus2330 commented on src/plugins/process/process.c:
```diff
@@ -0,0 +1,243 @@
+	// and this key should obviously contain the plugin's name, so check for any name
+	// furthermore by invoking process in a process we'd create a deadloop, check that too
+	const char * pluginName = keyString (pluginNameKey);
+	if (elektraStrCmp (pluginName, "(null)") == 0 || elektraStrCmp (pluginName, "(binary)") == 0)
```
Isn't keyIsBinary the better check?

markus2330 replied:
still not resolved

markus2330 commented:
#1817 was merged, please rebase and make this ready to merge.

e1528532 commented:
I've fixed the warning and noticed that if the key you use for the proxied plugins has an invalid name, it won't be appended to the keyset to be transferred to the child process, hence resulting in strange results and missing error messages. Fixed this issue now in general, so we have warning messages always.

However, the issue with the plugin configuration remains. Somehow it doesn't pass the configuration parameter to the plugin. I tried to print the whole configuration upon elektraProcessOpen by adding these few debug statements:

```
		KeySet * processConfig = elektraPluginGetConfig (handle);
		ksRewind (processConfig);
		Key * k;
		while ((k = ksNext (processConfig)) != NULL)
			ELEKTRA_LOG ("config got key %s %s", keyName (k), keyString (k));
```

However, when i mount a random file then using a proxied dump plugin like `sudo kdb mount test.dump user/examples/process process process=dump` the iteration doesn't print the /plugin key i'd expect to see what plugin should be proxied. The only message i get is about some `system/module` key in the plugin configuration: `src/plugins/process/process.c:64:elektraProcessOpen: config got key system/module this plugin was loaded for the status`
@markus2330 any idea why the configuration is not passed upon mounting?

markus2330 commented:
> any idea why the configuration is not passed upon mounting?

No, I am afraid you have to debug where the info is lost. Maybe it is not correctly passed to the other process and never reaches the proxied plugin? Please also add tests for this.

markus2330 commented:
Any progress here? I have yet another use case where this plugin would be handy. (For python)

e1528532 commented:
i see that you have already started to use the library directly for the python plugin. IMO for interpreters that always have to be executed inside an own process this is the better way than using this plugin as a proxy, because if it doesn't work without being executed in its own process it is of little use on its own.
I don't plan to continue on that before the typechecker and thesis is done due to the deadline

markus2330 commented:
Yes, I agree. Patching ruby and java to directly use the library also makes sense.

For other plugins, like xerces, however, it would be good to be able to choose if fork is used. So the plugin is still relevant. But yes, you need to concentrate on the type checker.

e1528532 commented:
rebased everything on the newest pluginprocess improvements, lets see if it helps

markus2330 commented:
Looks good. There only seems to be some issue if logging+debugging is turned on.

@ingwinlu Wouldn't it make sense to rename the -optimization to -log-debug? It is much more likely to trigger a bug in log+debug than to trigger a bug in the non-optimization case.

e1528532 commented:
Are they related to this plugin? Judging from the failed tests i don't see a direct connection to it, apart from check_kdb_internal_check:

```
	112 - testshell_markdown_ini (Failed)
	174 - testscr_check_kdb_internal_check (Failed)
	193 - testshell_replay_ls (Failed)
	199 - testshell_markdown_tutorial_cascading (Failed)
	200 - testshell_markdown_tutorial_validation (Failed)
	201 - testshell_markdown_msr_syntax (Failed)
	202 - testshell_markdown_readme_msr (Failed)
	203 - testshell_markdown_issue_template (Failed)
```

e1528532 commented:
Wait i see i accidentially committed log.c, maybe that causes it

e1528532 commented:
Finally found out that this plugin seems to be impossible with the current backend behavior.
Backend fetches the plugin's contract in `backendOpenModules` in `backend.c`, however it only provides a default config there for the plugin.

But this plugin needs to have its config available already, as it needs to grab the `/plugin` config in order to deliver the correct contract. The process plugin has to mimic the contract of the plugin it proxies, and for that i wanted to use `invoke` to call the get method of the proxied plugin with the contract key.

I've bypassed this for now by assuming there won't be a config available when grabbing the contract and returning a generic default, however this is most likely not what shall be done as this makes the information in `system/elektra/modules/process` wrong. The issue is that this information can't really be determined statically, as it totally depends on the plugin that is being proxied. Any idea how to work around this without tons of work? @markus2330

e1528532 commented:
In case there is no easy workaround i'd suggest merging this anyway after it passes the build in its current state, i've added this fallacy to the README. As the plugin is experimental i think its good to have it in master, maybe someone will continue on it one day and for simple things like proxying dump it works (to show the experiment)

markus2330 commented on src/plugins/process/process.c:
```diff
@@ -0,0 +1,286 @@
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		if (!validPluginName (pluginName, parentKey)) return ELEKTRA_PLUGIN_STATUS_SUCCESS;
```
Please comment what is done here?

Shouldn't the contract be renamed to pluginName?

e1528532 replied:
as said below, it seems to be expected to be called process

markus2330 commented on src/plugins/process/process.c:
```diff
@@ -0,0 +1,286 @@
+		{
+			keySetBaseName (cpy, NULL);
+			keySetBaseName (cpy, NULL);
+			keyAddBaseName (cpy, "process");
```
Why to process and not pluginName?

e1528532 replied:
the contract returned from this plugin seems to be expected to be called process, i don't recall the error but i think it was something like contract missing.

markus2330 replied:
Ok. Did you try to use `process#1 plugin=dump process#2 plugin=other`? Iirc this should allow different contracts for plugins with the same name.

e1528532 replied:
unfortunately this still doesn't resolve the issue of not knowing which plugin shall be proxied upon getting the contract, so it is only of little use currently.

markus2330 replied:
Ok, please resolve the other small things so that we can get this into master.

markus2330 commented on src/plugins/process/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## Limitations
+
+This plugin cannot act as a proxy for itself to prevent loops. Currently only plugins that
```
Make a list and turn around. The second limitation is much more important.

e1528532 replied:
done

markus2330 commented on src/plugins/process/README.md:
```diff
@@ -0,0 +1,52 @@
+The process plugin uses the `pluginprocess` library in order to execute other plugins
+inside an own process.
+This is useful for plugins which cause memory leaks to be isolated in an own process.
+Furthermore this is useful for runtimes or libraries that cannot be reinitialized in
```
Second reason more important (swap)

e1528532 replied:
done

markus2330 commented on src/plugins/process/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+				 elektra-invoke)
+
+else (PLUGINPROCESS_FOUND)
+	message ("${PLUGINPROCESS_NOTFOUND_INFO}, excluding pluginprocess library, thus excluding the process plugin")
```
pluginprocess library excluded, thus...

e1528532 replied:
done

markus2330 commented on src/plugins/process/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+	return ()
+endif (NOT_INCLUDED)
+
+if (PLUGINPROCESS_FOUND)
```
wrongly used, please properly exclude plugin

e1528532 replied:
done

markus2330 commented on src/error/specification:
```diff
@@ -1230,3 +1230,17 @@ severity:warning
 macro:ZEROMQSEND_ERROR
+
+number: 197
+description:Failed to initialize the execution of a plugin in a child process
```
reuse errors from pluginprocess?

e1528532 replied:
Yes i think that error can be reused to avoid introducing more as this plugin and pluginprocess are quite closely related

markus2330 commented on cmake/Modules/LibAddPlugin.cmake:
```diff
@@ -179,6 +179,7 @@ function (add_plugintest testname)
 endfunction (add_plugintest)
 
 function (plugin_check_if_included PLUGIN_SHORT_NAME)
+	set (NOT_INCLUDED "" PARENT_SCOPE)
```
What is done here?

markus2330 commented on src/plugins/process/README.md:
```diff
@@ -0,0 +1,52 @@
+- infos/provides =
+- infos/recommends =
+- infos/placements = pregetstorage getstorage postgetstorage presetstorage setstorage
+- infos/status = maintained unittest shelltest tested configurable experimental
```
status unfinished missing. Is it really tested?

markus2330 replied:
Yes, thank you. Better to understate.

markus2330 commented:
@e1528532  would be great to have this for the upcoming release which hopefully happens today.

Please also improve the release notes about the type checker plugin

markus2330 commented:
Thank you, the sooner the better :+1:

Actually only the update of the release notes is necessary. If the processplugin is within this release or the next does not really matter.

markus2330 commented:
The build fails now, let us merge this after the release.

e1528532 commented:
accidentially wrote exclude_plugin instead of remove_plugin so that one test fails. debian-unstable-clang failed due to zeromq,so i guess thats not my fault. I've corrected the exclude_plugin but since all the other tests ran successfully i don't think its necessary to wait for them to complete.

e1528532 commented:
adding ready to merge, judge for yourself if you trust my explanation ;)

markus2330 commented:
Thank you! Sorry it was too late, I already started the release when I saw your "ready to merge".

Please rebase for the release notes.

e1528532 commented:
jenkins build libelektra please

markus2330 commented:
Thank you!
